### DISCUSSION TITLE
How to force urql to give __typename always, even on cache hit return?

### DISCUSSION THREAD

**Discussion by RIP21** [urql-graphql/urql#1241] [Q&A] [answered]
When I open view that fetches some data and then `console.log` the `data`. I see that it has various requested fields + `__typename`
But if I move out of this view and return back to it, console.log shows that `__typename` field disappeared :) 
Is there any way to keep `__typename` in the `data` no matter when data just returned from the network or has been taken from cache :) 

It's just for the sake of consistency and I think this is what people expect.

The workaround I found is to explicitly request `__typename` :) I noticed that as I had some rendering logic that was depending on the `__typename` :) So I branched based on it :) 

Thanks, I'm using the latest urql + graphcache

**Comment by kitten** [comment, marked as answer]
It's normal that `urql` doesn't add `__typename` automatically to the root type of the document. In general you shouldn't rely on this and instead should assume that it merely _may_ be there _unless_ you explicitly ask for it.

That being said Graphcache and the document transform in `@urql/core` in general adds the `__typename` field to any selection set except for the root selection set on your root types.

**Reply by RIP21** [reply]
Makes sense :) Ok then :) Thanks :)

**Comment by wereHamster** [comment]
After upgrading to urql 2 and graphcache 4, I'm running into the same problem. The data includes `__typename` everywhere when the data is fetched fresh from the server, but on cache hit _none_ of the objects (root nor nested) have `__typename`.

I'm generating my queries via graphql-codegen. Do you know if there is a way to have it automatically add `__typename` everywhere? I've tried setting `nonOptionalTypename: true`, but that only affects the types, not the actual queries.

**Reply by kitten** [reply]
This is probably a good issue report to give to GCG. There's no automatic way to add the type name field to every document and it's not added automatically in cache-bound requests to only return requested fields as often as possible.

So, if you need the field, adding it explicitly will return it as expected

**Reply by wereHamster** [reply]
I gave the whole situation around __typename in GCG and urgl a thorough thought and am now of the opinion that I should explicitly add __typename to the queries where needed. The less magic there is, the better. I don't consider it good practice anymore to try to automatically add __typename to all queries  (through GCG, or another mechanism). In the vast majority of cases I don't need it, and where I need it I want to make it a conscious decision.

That being said, I find it slightly concerning that urql is returning different data in the first query vs. when it's coming from the cache. Wouldn't it be more consistent to always only return what the user asks for (ie. never include __typename if the user didn't ask for it)?

**Reply by kitten** [reply]
@wereHamster I think this comes down to us leaking some server information to your data in that not all data is being properly regenerated so I'll take a look and will see whether I can make the results more consistent once more 🙌

**Reply by dylan-chong** [reply]
> I find it slightly concerning that urql is returning different data in the first query vs. when it's coming from the cache.

this is big. because the cached version does not have __typename, it's triggering a deep set of rerenders (i'm doing deep comparison to check if the result has changed) and this is causing issues

**Reply by kitten** [reply]
@dylan-chong This is a 3yr old thread and the issues described in here have been resolved. Assuming you're on the latest version of Graphcache, if you run into unrelated or new issues please open a bug report or a new discussion thread ❤️
